### PR TITLE
Move onboarding emails to scheduled function

### PR DESCRIPTION
**Pull request by IanPhilips**
In order to respect users' email preferences we move from scheduling the emails on mailgun to a scheduled cloud function

Screenshot 2022-11-21 at 11 52 24 AM

### CODE REVIEW

**Comment by IanPhilips** [comment]
@mqp Do you have any idea how I could make a v2 function whose permissions only allow another cloud function to call it? E.g. what you have for the update user metrics but instead of no auth required, **only** allow a pubsub function to call the v2 function.

**Comment by mqp** [comment]
> @mqp Do you have any idea how I could make a v2 function whose permissions only allow another cloud function to call it? E.g. what you have for the update user metrics but instead of no auth required, **only** allow a pubsub function to call the v2 function.

The most obvious way I would implement this is to make a secret that only those two functions have access to, which the first seconds to the second, and the second checks.

File: functions/src/onboarding-emails.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  const day = dayjs().utc().day()
+  // don't run on sun or mon
+  if (day === 0 || day === 1) return
+  const contracts = await getTrendingContracts()
+  const { users, idToPrivateUsers } = await getUsersCreatedBetweenDays(1, 2)
+  await Promise.all(
+    users.map(async (user) => {
+      const privateUser = idToPrivateUsers.get(user.id)
+      // If it's a wednesday we might get users created on monday that've already gotten the weekly email
+      if (!privateUser || privateUser.weeklyTrendingEmailSent) return
```
Comment (mantikoros) [inline functions/src/onboarding-emails.ts:62]: I'm assuming weeklyTrendingEmailSent gets reset sometime.

File: functions/src/onboarding-emails.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  const maxDaysTime = dayjs().subtract(maxDaysAgo, 'days').valueOf()
+
+  console.log(
+    `Getting users created between ${minDaysAgo} and ${maxDaysAgo} days ago`
+  )
+  const users = (
+    await firestore
+      .collection('users')
+      .where('createdTime', '<', minDaysTime)
+      .where('createdTime', '>', maxDaysTime)
```
Comment (mantikoros) [inline functions/src/onboarding-emails.ts:109]: Did you make sure an firestore index was created for this on prod?

File: functions/src/onboarding-emails.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  const usersAndPrivateUsersMap = new Map<string, PrivateUser>()
+  await Promise.all(
+    users.map(async (user) => {
+      const privateUser = await firestore
+        .collection('private-users')
+        .doc(user.id)
+        .get()
+        .then((doc) => doc.data() as PrivateUser)
+      usersAndPrivateUsersMap.set(user.id, privateUser)
```
Comment (mantikoros) [inline functions/src/onboarding-emails.ts:121]: It's bad practice to put a mutable expression like usersAndPrivateUsersMap.set() in normally immutable function like map(). You could have the function return a [string, PrivateUser] and then construct an object from the entries instead.

**Review by mantikoros — APPROVED** [review]
The logic looks alright to me. And I suppose this does provide additional functionality (following user email preferences), but adding a new brittle process that loads all users into memory every day is a little less robust than the current implementation.

File: functions/src/onboarding-emails.ts (hunk 2 above)
Comment (IanPhilips) [reply functions/src/onboarding-emails.ts:109]: I will!

File: functions/src/onboarding-emails.ts (hunk 1 above)
Comment (IanPhilips) [reply functions/src/onboarding-emails.ts:62]: Yup, on monday mornings

**Comment by IanPhilips** [comment]
> > @mqp Do you have any idea how I could make a v2 function whose permissions only allow another cloud function to call it? E.g. what you have for the update user metrics but instead of no auth required, **only** allow a pubsub function to call the v2 function.
> 
> The most obvious way I would implement this is to make a secret that only those two functions have access to, which the first sends to the second, and the second checks.

Cool, that sounds reasonable & easy. Was thinking there could be some other way I was missing.

**Comment by IanPhilips** [comment]
One thought: can you un-schedule a mailgun email via email address?

**Comment by IanPhilips** [comment]
Oh nvm I looked this up previously, you can only dequeue them all

**Comment by IanPhilips** [comment]
Did this: 04932cae239f965ded986148394cd7c3f221d063

**Pull request closed without merging**